### PR TITLE
fix: concurrency insert between pending and delivered

### PR DESCRIPTION
**Pull request by Freyskeyd**
# Description

This PR is covering a corner case which involve the delivery of a prev certificate and a child certificate.

The scenario is this one:

- A `prev` certificate (`C1`) has been delivered (by the broadcast) and need to be persisted
  The persist method will hold a lock while performing multiple insert/delete to avoid
  insert race condition.
- At the same time, another node is sending a certificate (`C2`) which have `C1` as `prev_id`.
  `C2` is looking at the storage to find if the `prev_id` `C1` is delivered but find nothing as
  the `persist` method is still working at creating the `WriteBatch`. It led the node to put
  `C2` in the `precedence_pool` waiting for `C1` to be delivered while it is in fact already
  delivered.

To avoid that and as a first step, when trying to insert a certificate in the pending pool,
The node will try to acquire a lock guard on the certificate but also on the prev_id.

## PR Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added or updated tests that comprehensively prove my change is effective or that my feature works

### CODE REVIEW

File: crates/topos-tce-storage/src/tests/pending_certificates.rs [hunk 1]
```diff
@@ -84,5 +87,128 @@ async fn adding_pending_certificate_already_delivered(store: Arc<ValidatorStore>
 
     assert!(store
         .insert_pending_certificate(&initial_certificate_delivered.certificate)
+        .await
         .is_err());
 }
+
+/// This test is covering a corner case which involve the delivery of a prev certificate
```
Comment (dvdplm) [inline crates/topos-tce-storage/src/tests/pending_certificates.rs:94]:
```suggestion
/// This test is covering a corner case which involves the delivery of a prev certificate
```

File: crates/topos-tce-storage/src/validator/mod.rs [hunk 2]
```diff
@@ -286,10 +286,18 @@ impl ValidatorStore {
         Ok(ids)
     }
 
-    pub fn insert_pending_certificate(
+    pub async fn insert_pending_certificate(
         &self,
         certificate: &Certificate,
     ) -> Result<Option<PendingCertificateId>, StorageError> {
+        // A lock guard is asked during the insertion of a pending certificate
```
Comment (dvdplm) [inline crates/topos-tce-storage/src/validator/mod.rs:293]:
```suggestion
        // A lock guard is taken during the insertion of a pending certificate
```

File: crates/topos-tce-storage/src/validator/mod.rs [hunk 3]
```diff
@@ -286,10 +286,18 @@ impl ValidatorStore {
...
+        // to avoid race condition when a certificate is being inserted and added
+        // to the pending pool at the same time
```
Comment (dvdplm) [inline crates/topos-tce-storage/src/validator/mod.rs:295]:
```suggestion
        // to avoid race condition when a certificate (C1) is delivered and is in the process of being persisted to disk (so no longer pending), and a dependent certificate (C2) is added
        // to the pending pool even though it could be delivered as well.
```

File: crates/topos-tce-storage/src/validator/mod.rs [hunk 4]
```diff
@@ -312,6 +320,14 @@ impl ValidatorStore {
             ));
         }
 
+        // A lock guard is asked during the insertion of a pending certificate
```
Comment (dvdplm) [inline crates/topos-tce-storage/src/validator/mod.rs:323]:
```suggestion
        // A lock guard is taken during the insertion of a pending certificate
```

File: crates/topos-tce-broadcast/src/tests/task_manager.rs [hunk 5]
```diff
@@ -19,8 +19,11 @@ use crate::{sampler::SubscriptionsView, task_manager::TaskManager};
 
 #[rstest]
 #[tokio::test]
-async fn can_start(#[future] create_validator_store: Arc<ValidatorStore>) {
-    let validator_store = create_validator_store.await;
+async fn can_start(
+    #[future(awt)]
```
Comment (dvdplm) [inline crates/topos-tce-broadcast/src/tests/task_manager.rs:23]: What does this do?

**Review by dvdplm — APPROVED** [review]
LGTM, and thank you for the high quality comments explaining what is going on.

Beyond this PR, I think this need for strong synchronization is unfortunate. It makes the code less clear and introduces delays that could harm performance. Not sure how we could do better but I'll try to give it some thought. I have a gut feeling that treating pending and delivered certificates separately is part of the problem, but I'll have to think more on this.

**Review by gruberb — APPROVED** [review]

File: crates/topos-tce-broadcast/src/tests/task.rs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    let validator_id = ValidatorId::default();
+    let thresholds = topos_config::tce::broadcast::ReliableBroadcastParams {
+        echo_threshold: 1,
+        ready_threshold: 1,
+        delivery_threshold: 1,
+    };
+    let (event_sender, mut event_receiver) = mpsc::channel(2);
+    let (broadcast_sender, _) = broadcast::channel(1);
+    let message_signer = Arc::new(
+        MessageSigner::from_str("122f3ae6ade1fd136b292cea4f6243c7811160352c8821528547a1fe7c459daf")
```
Comment (atanmarko) [inline crates/topos-tce-broadcast/src/tests/task.rs:44]: Could you put it in `topos_test_sdk::constants`?

**Review by atanmarko — APPROVED** [review]

File: crates/topos-tce-storage/src/validator/mod.rs (hunk 3 above)
Comment (Freyskeyd) [reply crates/topos-tce-storage/src/validator/mod.rs:295]: That's not quit true but I will rephrase

File: crates/topos-tce-broadcast/src/tests/task_manager.rs (hunk 5 above)
Comment (Freyskeyd) [reply crates/topos-tce-broadcast/src/tests/task_manager.rs:23]: it awaits the future for us

File: crates/topos-tce-storage/src/fullnode/mod.rs [hunk 7]
```diff
@@ -82,6 +83,27 @@ impl FullNodeStore {
             index_tables,
         }))
     }
+
+    /// Await for a [`LockGuards`] for the given certificate id
+    pub(crate) async fn get_certificate_lock_guard(
```
[the remark this answers is not in the document]
Comment (Freyskeyd) [reply crates/topos-tce-storage/src/fullnode/mod.rs:88]: For me we want to get a `lock_guard` for a particular certificate id from what's close to a `HashMap`, what are your arguments not to use `get_` here?

[the remark this answers is not in the document]
Comment (dvdplm) [reply crates/topos-tce-storage/src/fullnode/mod.rs:88]: "get" is implied, you don't really need it. And it's not idiomatic Rust (see the [API guidelines](https://rust-lang.github.io/api-guidelines/naming.html#c-getter)).
